### PR TITLE
Widget strucured classification

### PR DESCRIPTION
**Pull request by mishig25**
@osanseviero please let me know how it looks.

#### Columns
Specifically, the columns part we discussed [here](https://github.com/huggingface/huggingface_hub/blob/b87caab21b56c6cbfd7f16caca849749ca5af488/widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte#L33), [here](https://github.com/huggingface/huggingface_hub/blob/b87caab21b56c6cbfd7f16caca849749ca5af488/widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte#L46) & [here](https://github.com/huggingface/huggingface_hub/blob/b87caab21b56c6cbfd7f16caca849749ca5af488/widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte#L68)

#### API Contract:
input: 
```js
const requestBody = {
    inputs: {
        data: convertTableToData(table),
    },
};
```
output: identical to TableQAWidget

### CODE REVIEW

File: widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte
Comment (osanseviero) [inline widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte:152]: I think we should refactor this. It's the same as Table QA no? You could put them 
in https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/helpers.ts

File: widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte
Comment (osanseviero) [inline widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte:186]: I'm not sure this will work for the output. The output of the Inference API will just be a list of numbers (which could be a new added column). Same with isValidOutput. Feel free to check https://github.com/huggingface/huggingface_hub/blob/main/api-inference-community/docker_images/sklearn/app/pipelines/structured_data_classification.py#L11

File: widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte
Comment (mishig25) [reply widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte:152]: totally! should I put them [helpers.ts](https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/helpers.ts) or [ViewUtils.ts](https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/ViewUtils.ts)?

Comment (osanseviero) [reply widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte:152]: Hmm ViewUtils seems like a better place

**Comment by mishig25** [comment]
@osanseviero updates:
* `output` gets added as a new column
* `output` value/row gets colored based on its value/class
* supports dark mode

todo:
* more robust reset function
* implement `remove` functionalities for row * column
* better looking colors 
* test against deployed pipeline (when I try against `julien-c/wine-quality`, I get `Pipeline is not defined for model julien-c/wine-quality`, please let me when there is a deployed pipeline for structured data)

**Comment by osanseviero** [comment]
Feel free to let me know when this is ready for review

**Comment by mishig25** [comment]
@osanseviero ready for review

File: widgets/src/lib/InferenceWidget/shared/ViewUtils.ts
Comment (osanseviero) [inline widgets/src/lib/InferenceWidget/shared/ViewUtils.ts:104]: Nit: indentation seem wrong

File: widgets/src/lib/InferenceWidget/shared/helpers.ts
Comment (osanseviero) [inline widgets/src/lib/InferenceWidget/shared/helpers.ts:144]: nit: remove to reduce unneeded diff

File: widgets/src/lib/InferenceWidget/shared/WidgetTableInput/WidgetTableInput.svelte
Comment (osanseviero) [inline widgets/src/lib/InferenceWidget/shared/WidgetTableInput/WidgetTableInput.svelte:34]: nit: remove to avoid unneeded diff

**Review by osanseviero — APPROVED** [review]
This looks awesome! 👍 

One small nit: same way you scroll to the predictions column, we could scroll to the missing input cell.

I'll let @julien-c review the code a bit more, but overall this is great!

**Comment by julien-c** [comment]
i'm not mega familiar with the widgets codebase anymore so I'll let @gary149 and/or @beurkinger take a quick look at this instead!

@mishig25 mind uploading a screen capture of the UX to this PR and/or the associated issue?

**Comment by osanseviero** [comment]
Btw you can try it out directly in https://60ec6ac95a2e0d37d3d6e172--huggingface-widgets.netlify.app/ (search Structured Data Classification)

**Comment by julien-c** [comment]
Oh yeah awesome!

**Comment by julien-c** [comment]
How do you feel about Predictions popping as a final column @gary149?

**Comment by mishig25** [comment]
@julien-c please let me know if you still need screen capture (I think the link Omar supplied did the job). 
@gary149 Omar & I settled on the predictions popping up as the last column. Let me know Predictions column should be presented differently

File: widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte
Comment (mishig25) [inline widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte:70]: @gary149 here is the list of `colors` for the Predictions. Please let me know if you have suggestions on the list.
More details:
```
let arr_unique = unique([1,1,2,2,3,3]) --> [1,2,3]
If arr_unique.length <= colors.length:
    assign color to every unique class --> all 1s are same blue, all 2s are green, etc.
else:
  assign color by iterating colors --> first element blue, second element green, nth element colors[mod(n, color.length)]
```

**Comment by mishig25** [comment]
@julien-c since both Thibault and Victor are on vacation, I've carefully re-reviewed myself this MR and believe its ready to be merged. If you give your thumbs up, I'll merge it on hub & internally. 
strucured-data-classification is already online, so this widget should just work once merged. Please let me know 👍

**Comment by julien-c** [comment]
Only gave it a quick glance but LGTM, feel free to merge and deploy :)

**Pull request merged**
